File: src/p0054/golang/solution.go
/**
 * Given an m x n matrix, return all elements of the matrix in spiral order.
 * 
 *  
 * 
 * Example 1:
 * 
 * 
 * Input: matrix = [[1,2,3],[4,5,6],[7,8,9]]
 * Output: [1,2,3,6,9,8,7,4,5]
 * Example 2:
 * 
 * 
 * Input: matrix = [[1,2,3,4],[5,6,7,8],[9,10,11,12]]
 * Output: [1,2,3,4,8,12,11,10,9,5,6,7]
 *  
 * 
 * Constraints:
 * 
 * m == matrix.length
 * n == matrix[i].length
 * 1 <= m, n <= 10
 * -100 <= matrix[i][j] <= 100
 */

func spiralOrder(matrix [][]int) []int {
	m, n := len(matrix), len(matrix[0])
	ans := make([]int, 0, m*n)
	left, top, right, bottom := 0, 0, n-1, m-1
	for left <= right && top <= bottom {
		for i := left; i <= right; i++ {
			ans = append(ans, matrix[top][i])
		}
		for i := top + 1; i <= bottom; i++ {
			ans = append(ans, matrix[i][right])
		}
		if left < right && top < bottom {
			for i := right-1; i > left; i-- {
				ans = append(ans, matrix[bottom][i])
			}
			for i := bottom; i > top; i-- {
				ans = append(ans, matrix[i][left])
			}
		}
		left++
		top++
		right--
		bottom--
	}
	return ans
}

File: src/p0054/python/solution.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        m, n = len(matrix), len(matrix[0])
        visited = set()
        ans = []
        i, j = 0, 0
        while len(ans) < m*n:
            while j < n and (i, j) not in visited:
                ans.append(matrix[i][j])
                visited.add((i, j))
                j += 1
            i, j = i + 1, j - 1
            while i < m and (i, j) not in visited:
                ans.append(matrix[i][j])
                visited.add((i, j))
                i += 1
            i, j = i - 1, j - 1
            while j >= 0 and (i, j) not in visited:
                ans.append(matrix[i][j])
                visited.add((i, j))
                j -= 1
            i, j = i - 1, j + 1
            while i >= 0 and (i, j) not in visited:
                ans.append(matrix[i][j])
                visited.add((i, j))
                i -= 1
            i, j = i + 1, j + 1
        
        return ans
    
# 华为